src/Swift.Bindings/src/Emitter/StringEmitter/Handler/MethodHandler.cs

```diff
@@ -589,28 +639,45 @@ private void EmitGenericArguments(IndentedTextWriter writer)
                 {{payloadName}} = (IntPtr)NativeMemory.Alloc({{metadataName}}.Size);
                 SwiftMarshal.MarshalToSwift({{argument.Name}}, {{payloadName}});
                 """;
-                writer.WriteLines(text);
+                csWriter.WriteLines(text);
             }
-            writer.WriteLine();
+            csWriter.WriteLine();
         }
 
         /// <summary>
         /// Emits the PInvoke call.
         /// </summary>
         /// <param name="writer">The IndentedTextWriter instance.</param>
-        private void EmitPInvokeCall(IndentedTextWriter writer)
-        {
-            var voidReturn = _env.MethodDecl.CSSignature.First().SwiftTypeSpec.IsEmptyTuple;
-            var returnPrefix = (_requiresIndirectResult || voidReturn) ? "" : "var result = ";
-            writer.WriteLine($"{returnPrefix}{NameProvider.GetPInvokeName(_env.MethodDecl)}({_pInvokeSignature.CallArgumentsString()});");
-            writer.WriteLine();
+        private void EmitPInvokeCall(CSharpWriter csWriter)
+        {
+            if (_requiresSwiftAsync)
+            {
+                string parameters = string.Join(", ",
+                    new[]
+                    {
+                        $"(IntPtr)s_{_env.MethodDecl.Name}Callback",
+                        "IntPtr.Zero",
+                        "GCHandle.ToIntPtr(handle)",
+                        _pInvokeSignature.CallArgumentsString()
+                    }.Where(arg => !string.IsNullOrEmpty(arg))
```

Comment:
I think this can be moved to Signature Builder. So that `_pInvokeSignature.CallArgumentsString()` always returns a correct list.

Reply:
Added specific handler in the PInvokeSignature.